File: y2018/day19/day19.py
import re
from collections import defaultdict
from y2018.day16.ops import Op, op_res

P2_RUN_AMT = 1000

ip_fmt = r'#ip ([0-5])'
instruction_fmt = r'([a-zA-Z]+) (\d+) (\d+) (\d+)'
instructions = []
with open('day19.txt') as read:
    lines = read.readlines()
    ip = int(next(iter(re.findall(ip_fmt, lines[0]))))
    for l in lines[1:]:
        op, a, b, c = next(iter(re.findall(instruction_fmt, l)))
        a, b, c = map(int, (a, b, c))
        instructions.append((Op[op.upper()], a, b, c))

ip_val = 0
registers = defaultdict(int)
while 0 <= ip_val < len(instructions):
    i = instructions[ip_val]
    registers[ip] = ip_val
    registers = op_res(registers, i[1:], i[0])
    ip_val = registers[ip]
    ip_val += 1

print(f"p1 register 0 val: {registers[0]}")

ip_val = 0
registers = defaultdict(int)
registers[0] = 1
for _ in range(P2_RUN_AMT):
    i = instructions[ip_val]
    registers[ip] = ip_val
    registers = op_res(registers, i[1:], i[0])
    ip_val = registers[ip]
    ip_val += 1

# explanation sauce: https://todd.ginsberg.com/post/advent-of-code/2018/day19/
giant_num = max(registers.values())
total = 0
for i in range(1, giant_num + 1):
    if giant_num % i == 0:
        total += i
print(f"p2 register 0 val: {total}")

File: y2018/rust/src/bin/day19.rs
use aoc_rust_2018::assembly::Op;
use regex::Regex;
use std::fs;

const P2_RUN_AMT: u32 = 100;

fn main() {
    // YOU'RE ADOPTED, RUST.
    let read = fs::read_to_string("../input/day19.txt").unwrap();
    let mut lines = read.lines();
    let ip_fmt = Regex::new("#ip ([0-5])").unwrap();
    let instruct_fmt = Regex::new(r"([a-zA-Z]+) (\d+) (\d+) (\d+)").unwrap();
    let ip = ip_fmt.captures(lines.next().unwrap()).unwrap()[1].parse::<usize>().unwrap();
    let mut instructions = Vec::new();
    for i in lines {
        if let Some(m) = instruct_fmt.captures(i) {
            instructions.push((
                Op::from_str(&m[1]).unwrap(),
                m[2].parse::<i32>().unwrap(),
                m[3].parse::<i32>().unwrap(),
                m[4].parse::<i32>().unwrap(),
            ));
        }
    }

    let mut ip_val: i32 = 0;
    let mut registers = vec![0; 6];
    while 0 <= ip_val && ip_val < instructions.len() as i32 {
        let (op, a, b, c) = instructions[ip_val as usize];
        registers[ip] = ip_val;
        registers = op.apply(registers.clone(), a, b, c);
        ip_val = registers[ip];
        ip_val += 1;
    }

    println!("p1 register 0 val: {}", registers[0]);

    ip_val = 0;
    registers = vec![0; 6];
    registers[0] = 1;
    for _ in 0..P2_RUN_AMT {
        // yeah, this code is duplicated. unfortunate, really
        let (op, a, b, c) = instructions[ip_val as usize];
        registers[ip] = ip_val;
        registers = op.apply(registers.clone(), a, b, c);
        ip_val = registers[ip];
        ip_val += 1;
    }

    let giant_num = *registers.iter().max().unwrap();
    let mut total = 0;
    for i in 1..=giant_num {
        if giant_num % i == 0 {
            total += i;
        }
    }

    println!("p2 register 0 val: {total}");
}
